### PR TITLE
Rename SigHash to Sighash

### PR DESCRIPTION
**Pull request by tcharding**
Our usage of `SigHash` implies that 'sighash' is _two_ words; 'sighash' is a well known word in the Bitcoin ecosystem it should appear in identifiers as `Sighash`.

Change various types, variants, and code comments to use sighash as a single word. 

- Patches 1-8 are code changes `s/SigHash/Sighash/g`
- Patch 9 is code changes `s/sig_hash/sighash/g`
- Patch 11 is docs fixes

Fixes: #911

## Note to reviewers

I've been particularly pedantic with the patch separation because we are so close to release.

Done as separate patches to make review easier if review is to be done by reading the diffs. Perhaps at least one person could verify this PR programmatically by doing
- Reset the last 2 patches (those are easy to do manually)
- Check out master
- Do `s/SigHash/Sighash/g` on all source files (bash function below)
- Use `git diff branchA..branchB` to verify

The difference between the two branches should only include comment lines (last three patches) and these seven instances of `SigHash:

```
CHANGELOG.md:82:- [Add FromStr/Display implementation for SigHashType](https://github.com/rust-bitcoin/rust-bitcoin/pull/497/commits/a4a7035a947998c8d0d69dab206e97253fd8e048)
CHANGELOG.md:93:- [Introduce `SigHashCache` structure](https://github.com/rust-bitcoin/rust-bitcoin/pull/390) to replace `SighashComponents` and support all sighash modes
CHANGELOG.md:121:    - `SigHash`
src/blockdata/transaction.rs:1190:            "SigHash_None",
src/blockdata/transaction.rs:1191:            "SigHash_NONE",
src/util/sighash.rs:1175:            "SigHash_None",
src/util/sighash.rs:1176:            "SigHash_NONE",
```

In case its useful, the shell function I used to do these changes is:
```bash
function search-and-replace() {
    if (($# != 2))
    then
        echo "Usage: $0 <this> <that>"
        return
    fi

    local this="$1"
    local that="$2"

    # For all files containing $this, replace $this with $that.
    for file in $(git grep -l "$this")
    do
        perl -pi -e "s/$this/$that/g" "$file"
    done
}
```

### CODE REVIEW

**Comment by tcharding** [comment]
Bother, I've forgotten to do `sig_hash` in variable names.

**Comment by thomaseizinger** [comment]
> Our usage of `SigHash` implies that 'sighash' is _two_ words

But it is two words, isn't it? It is the short form of "Signature Hash". Not saying I disagree with the rename (perhaps Core has a convention here that we can follow), just pointing out that the current naming is not completely unreasonable :)

**Comment by tcharding** [comment]
Check out the discussion thread on the resolved issue :)

**Comment by thomaseizinger** [comment]
> > But it is two words, isn't it? It is the short form of "Signature Hash".
> 
> Check out the discussion thread on the resolved issue :)

Ignore me then! I stopped reading the issue text after the shell-script because I wasn't interested in that and missed the fact that this "fixes" an issue. My bad!

**Comment by tcharding** [comment]
> this "fixes" an issue

No, my bad. I'll put that at the top. I had to choose Wall Of Text vs massive unexplained diff right at the end of the release cycle, sorry for making you guys read so much :)

**Comment by tcharding** [comment]
My dev self improvement focus at the moment is not writing many words when a few will do - its hard :)

**Comment by sanket1729** [comment]
I don't have a strong preference for `SighHash` vs `Sighash`. But I like that we are uniform with this.

File: src/util/bip143.rs [hunk 1]
```diff
@@ -107,19 +107,19 @@ impl SighashComponents {
 }
 
 /// A replacement for SigHashComponents which supports all sighash modes
-#[deprecated(since = "0.27.0", note = "please use [sighash::SigHashCache] instead")]
-pub struct SigHashCache<R: Deref<Target = Transaction>> {
-    cache: sighash::SigHashCache<R>,
+#[deprecated(since = "0.27.0", note = "please use [sighash::SighashCache] instead")]
+pub struct SighashCache<R: Deref<Target = Transaction>> {
```
Comment (sanket1729) [inline src/util/bip143.rs:111]: I think we should revert this change. It does not make sense to rename something that is deprecated.

File: src/blockdata/transaction.rs [hunk 2]
```diff
@@ -737,7 +737,7 @@ impl error::Error for NonStandardSighashType {}
 
 /// Legacy Hashtype of an input's signature
 #[deprecated(since = "0.28.0", note = "Please use [`EcdsaSighashType`] instead")]
-pub type SigHashType = EcdsaSighashType;
+pub type SighashType = EcdsaSighashType;
```
Comment (sanket1729) [inline src/blockdata/transaction.rs:740]: Same note here

**Review by sanket1729 — COMMENTED** [review]
Left two comments.

File: src/util/bip143.rs (hunk 1 above)
Comment (tcharding) [reply src/util/bip143.rs:111]: ooo yes, my bad. Thanks!

File: src/blockdata/transaction.rs (hunk 2 above)
Comment (tcharding) [reply src/blockdata/transaction.rs:740]: Will do, thanks.

File: src/util/bip143.rs (hunk 1 above)
Comment (tcharding) [reply src/util/bip143.rs:111]: Removed patch: `eaf2040 Rename SigHashType -> SighashType`, win for separate patches!

File: src/blockdata/transaction.rs (hunk 2 above)
Comment (tcharding) [reply src/blockdata/transaction.rs:740]: Removed patch: `eaf2040 Rename SigHashType -> SighashType`, win for separate patches!

File: src/util/bip143.rs (hunk 1 above)
Comment (tcharding) [reply src/util/bip143.rs]: Fixed `SigHashCache` change to only do the non-deprecated type (the one is `sighash` module).

**Comment by tcharding** [comment]
Force push has @sanket1729's changes in it.

**Review by sanket1729 — DISMISSED** [review]
ACK aa6717c9493a8ea5b097fcad9a32d44c88a1657c

**Comment by sanket1729** [comment]
The compiler is helpful in suggesting related names.

```
  --> src/descriptor/mod.rs:836:36
    |
836 |     use bitcoin::{self, secp256k1, EcdsaSigHashType, PublicKey};
    |                                    ^^^^^^^^^^^^^^^^
    |                                    |
    |                                    no `EcdsaSigHashType` in the root
    |                                    help: a similar name exists in the module: `EcdsaSighashType`

error[E0432]: unresolved import `bitcoin::util::sighash::SigHashCache`
  --> src/psbt/mod.rs:29:5
   |
29 | use bitcoin::util::sighash::SigHashCache;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^------------
   |     |                       |
   |     |                       help: a similar name exists in the module: `SighashCache`
   |     no `SigHashCache` in `util::sighash`

error[E0432]: unresolved import `bitcoin::SchnorrSigHashType`
  --> src/psbt/mod.rs:30:21
   |
30 | use bitcoin::{self, SchnorrSigHashType};
   |                     ^^^^^^^^^^^^^^^^^^
   |                     |
   |                     no `SchnorrSigHashType` in the root
   |                     help: a similar name exists in the module: `SchnorrSighashType`

```

**Comment by dr-orlovsky** [comment]
requires rebase now

**Review by apoelstra — DISMISSED** [review]
ACK aa6717c9493a8ea5b097fcad9a32d44c88a1657c

**Comment by dr-orlovsky** [comment]
needs rebase :(

**Comment by tcharding** [comment]
Changes in rebase:
- rebase on master which lead to three of the patches being empty

I have updated the PR description.

**Review by dr-orlovsky — APPROVED** [review]
ACK 46c34b3fb7530b7eeb7720efa38ebac6539a8b00

**Review by apoelstra — APPROVED** [review]
ACK 46c34b3fb7530b7eeb7720efa38ebac6539a8b00

I am going to merge this but it might cause rebasing issues.
